### ISSUE TITLE
System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')'

### ISSUE THREAD

**Issue by Arash-Sabet** [FluentValidation/FluentValidation#1698]
### FluentValidation version

10.0.2

### ASP.NET version

.net core 3.1

### Summary

The fluent validation version 9.5.4 used to be able to validate our model successfully. Upgrading to 10.0.2 results to a failure per the exception in the title of this issue. 

The stack trace is per below:

>   at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
>    at System.Collections.Generic.List`1.RemoveAt(Int32 index)
>    at FluentValidation.Internal.CollectionPropertyRule`2.GetValidatorsToExecute(ValidationContext`1 context) in C:\Projects\FluentValidation\src\FluentValidation\Internal\CollectionPropertyRule.cs:line 335
>    at FluentValidation.Internal.CollectionPropertyRule`2.FluentValidation.IValidationRuleInternal<T>.Validate(ValidationContext`1 context) in C:\Projects\FluentValidation\src\FluentValidation\Internal\CollectionPropertyRule.cs:line 125
>    at FluentValidation.AbstractValidator`1.Validate(ValidationContext`1 context) in C:\Projects\FluentValidation\src\FluentValidation\AbstractValidator.cs:line 92
>    at FluentValidation.AbstractValidator`1.Validate(T instance) in C:\Projects\FluentValidation\src\FluentValidation\AbstractValidator.cs:line 62

### Steps to Reproduce

The model that used to be validated successfully by FluentValidation ver 9.5.4 looks like the following code snippet:

```csharp
public class MyModel
{
		public dynamic Data { set => JData = value; }
		public JObject JData{ get; private set; }
		public bool Enable{ get; set; } = false;
		public string Id{ get; set; }
                public string[] OtherIds { get; set; }
}
```
Version 10.0.2 fails to validate this model. The validator's code is per below:

```csharp
public class MyModelValidator  : AbstractValidator<MyModel>
{
        public MyModelValidator()
        {
            RuleFor(m=> m.JData).NotNull();
            RuleForEach(m=> m.OtherIds)
                .NotEmpty()
                .MaximumLength(100)
                .When(m=> m.OtherIds?.Any() == true);
            RuleFor(m=> m.OtherIds)
                .Must(m=> m.Count() == m.Distinct().Count())
                .When(m=> m.OtherIds?.Any() == true);
        }
}
```

**Comment by JeremySkinner** [comment]
Thanks, yes this is a bug. I've fixed the issue and pushed out 10.0.3 to nuget. Please give it a try and let me know if you run into any further issues.

**Comment by Arash-Sabet** [comment]
Thanks, @JeremySkinner for the prompt answer. The issue is now fixed. Please close this ticket.
